psycopg/psycopg/connection.py

```diff
@@ -851,6 +989,32 @@ def notifies(self) -> Generator[Notify, None, None]:
                 n = Notify(pgn.relname.decode(enc), pgn.extra.decode(enc), pgn.be_pid)
                 yield n
 
+    @contextmanager
+    def pipeline(self) -> Iterator[None]:
+        """Context manager to switch the connection into pipeline mode."""
+        if self._pipeline is not None:
+            raise e.ProgrammingError("already in pipeline mode")
```

Comment:
Is it necessary to make pipeline non-reentering? I'm playing with a trivial optimisation of executemany for #145 by just wrapping the `cursor.executemany()` in a `conn.pipeline()` and it would be made more complex by checking whether we are or not in a pipeline already. How about https://github.com/psycopg/psycopg/commit/2f50a415ff7543329e01045e6b8ac4082df53761 ?

Reply:
Yes; looks reasonable to me, thanks.

Reply:
Playing with executemany tests, yesterday, I have noticed that the behaviour changes if we run:
```
cur.executemany()  # calls pipeline internally
cur.rowcount  # has received all the results
```
or
```
with conn.pipeline()
    cur.executemany()  # internal pipeline is no-o[
    cur.rowcount  # results are not returned
```
I would have expected that [triggering a flush](https://github.com/psycopg/psycopg/commit/a7ab0fdc615084846f0c65b86a2c2895f0f69488#diff-38c84614fefee615d2c96ae4cb1fec9e483f82ffc1d9606040f91e5fa8f2c809R237) would have made the results reach the cursor. Am I wrong expecting that? Do you think it's possible to reach that "state"? If not, maybe making `pipeline()` transparent to re-enter is not such a good idea.

Reply:
> I would have expected that [triggering a flush](https://github.com/psycopg/psycopg/commit/a7ab0fdc615084846f0c65b86a2c2895f0f69488#diff-38c84614fefee615d2c96ae4cb1fec9e483f82ffc1d9606040f91e5fa8f2c809R237) would have made the results reach the cursor. Am I wrong expecting that? Do you think it's possible to reach that "state"?

The flush tells the server that it can start returning result from now on, but the client still has to explicitly fetch those (calling `PQgetResult`). At the end of the pipeline, we fetch all results pending; this probably explains the difference you observed depending on whether the pipeline has exited or not when querying `.rowcount`. Maybe calling `cur.fetch*()` helps?